daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala

```diff
@@ -905,13 +907,26 @@ object Main extends Logging {
                     }
                   } else {
                     // not streaming, show left over data warning
+                    val Dump = new DataDumper
+                    var (dataText, dataHex) = ("", "")
+                    var curPositionInByte0b = inStream.inputSource.position
+                    var bytesAvailable = inStream.inputSource.bytesAvailable
+                    var bytesLimit = math.min(8, bytesAvailable).toInt
+                    val destArray = new Array[Byte](bytesLimit)
+                    val ret = inStream.inputSource.get(destArray, 0, bytesLimit)
```

Comment:
There are cases where this output might be misleading, primarly related when the left over bits do not start on a byte boundary. inputSource.get will get the raw underlying bytes, without regard to the bit offset. So if data ends mid-bit this is going to include those already read bits. I'm not sure of the best way to handle this. We doul use inStream.fillByteArray instead, which does take into account bits, but that stiill might be confusing showing full bytes that were offset by some amount of bits. But maybe this isn't a big deal?

Reply:
I think since the first line of the output informs the user about the consumed bits, the fact that the dump might include those bits might be ok. This dump is intended as a very rudimentary pointer at a possible issue in the data. The user could (and should) further analyze the issue if it isn't readily apparent from the dump, i think.

Reply:
It is an issue, and we debug a lot of data like this.

If we were going to try to address this, we could add to the dump a first-byte-of-dump key, which is only presented if the data ends mid-byte somwhere. It would be an individual-bits picture of the first byte of the left-over data. 
If say, the data ended at bit 5, and bit order is mostSignificantBitFirst, then this picture would look like xxxxx000 (where the 0's are the actual left over data bits in that first byte of the left over data.) If bitOrder is leastSignificantBitFirst the picture would look like 000xxxxx. 
I suggest this sort of enhancement would be best as a separate JIRA ticket however, as this is not thought through.

Reply:
Created [DAFFODIL- 2173](https://issues.apache.org/jira/browse/DAFFODIL-2173 ) to account for this issue